File: test/scripts/15-memory-access-patterns/correct.py
#!/usr/bin/env python3
"""
CORRECT: Sequential access - prefetcher helps.
Resume: "Optimize Memory Access Patterns"
"""
import signal
import time
N = 2_000_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def main():
    global stop
    data = list(range(N))
    print("CORRECT: Sequential access")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    start = time.perf_counter()
    laps = 0
    total = 0
    while not stop and (time.perf_counter() - start) < DURATION_SEC:
        for x in data:
            total += x
        laps += 1
    print(f"Stopped: {laps} laps, total={total}")

if __name__ == "__main__":
    main()

File: test/scripts/15-memory-access-patterns/correct.js
const N = 2_000_000;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const data = Array.from({ length: N }, (_, i) => i);
let laps = 0, total = 0;
const start = Date.now();
console.log('CORRECT: Sequential access');
const id = setInterval(() => {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    clearInterval(id);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  for (let i = 0; i < data.length; i++) total += data[i];
  laps++;
}, 0);
process.on('SIGINT', () => { clearInterval(id); process.exit(0); });
